p2p/channel.go

```diff
@@ -18,106 +18,184 @@
 	}
-	go c.serve(ln)
+
+	go c.readLoop()
+	go c.sendLoop()
+	go c.sendKeepaliveLoop()
+	c.Handle(topicKeepAlive, func(c Context) error {
+		log.Debug().Msg("Received P2P keep alive ping")
+		return c.OK()
+	})
 
 	return c, nil
 }
 
-func (c *Channel) listen() (*kcp.Listener, error) {
-	addr := fmt.Sprintf(":%d", c.listenPort)
-	ln, err := kcp.ListenWithOptions(addr, c.blockCrypt, 10, 3)
-	if err != nil {
-		return nil, fmt.Errorf("could not create p2p listener: %w", err)
+// readLoop reads incoming requests or replies to initiated requests.
+func (c *Channel) readLoop() {
+	for {
+		select {
+		case <-c.stopReadLoop:
+			return
+		default:
+			var msg transportMsg
+			if err := msg.readFrom(c.conn); err != nil {
+				continue
+			}
+
+			// If message contains topic it means that peer is making a request
+			// and waits for response.
+			if msg.topic != "" {
+				go c.handleRequest(&msg)
+			} else {
+				// In other case we treat it as a reply for peer to our request.
+				go c.handleReply(&msg)
+			}
+		}
 	}
-	return ln, nil
 }
 
-func (c *Channel) serve(ln *kcp.Listener) error {
-	// Configure server to use h2c.
-	h2s := &http2.Server{}
-	server := &http.Server{
-		Handler:      h2c.NewHandler(&requestsHandler{ch: c}, h2s),
-		ReadTimeout:  30 * time.Second,
-		WriteTimeout: 30 * time.Second,
+// sendLoop sends data to underlying network.
+func (c *Channel) sendLoop() {
+	for {
+		select {
+		case <-c.stopSendLoop:
+			return
+		case msg, more := <-c.sendQueue:
+			if !more {
+				return
+			}
+
+			if err := msg.writeTo(c.conn); err != nil {
+				log.Err(err).Msg("Write failed")
+			}
+		}
 	}
-	return server.Serve(ln)
+}
+
+// handleReply forwards reply message to associated stream result channel.
+func (c *Channel) handleReply(msg *transportMsg) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	if s, ok := c.streams[msg.id]; ok {
+		s.resCh <- msg
+	} else {
+		log.Warn().Msgf("Stream %d not found", msg.id)
+	}
+}
+
+// handleRequest handles incoming request and schedules reply to send queue.
+func (c *Channel) handleRequest(msg *transportMsg) {
+	c.mu.RLock()
+	handler, ok := c.handlers[msg.topic]
+	c.mu.RUnlock()
+
+	var resMsg transportMsg
+	resMsg.id = msg.id
+
+	if !ok {
+		resMsg.statusCode = statusCodePublicErr
+		errMsg := fmt.Sprintf("handler %q is not registered", msg.topic)
+		log.Err(errors.New(errMsg))
+		resMsg.data = []byte(errMsg)
+		c.sendQueue <- &resMsg
+		return
+	}
+
+	ctx := defaultContext{req: &Message{Data: msg.data}}
+	err := handler(&ctx)
+	if err != nil {
+		resMsg.statusCode = statusCodeInternalErr
+	} else if ctx.publicError != nil {
+		resMsg.statusCode = statusCodePublicErr
+		resMsg.data = []byte(ctx.publicError.Error())
+	} else {
+		resMsg.statusCode = statusCodeOK
+		if ctx.res != nil {
+			resMsg.data = ctx.res.Data
+		}
+	}
+	c.sendQueue <- &resMsg
+}
+
+// ServiceConn returns UDP connection which can be used for services.
+func (c *Channel) ServiceConn() *net.UDPConn {
```

Comment:
a bit unclear why we have this method here? I thought traversal package should handle this.

Reply:
Yes, traversal package does actual pinging. This field is not used yet, not sure yet how will second conn be passed to actual service, will remove it in separate PR.